core/trino-main/src/test/java/io/trino/operator/aggregation/TestDecimalAverageAggregation.java

```diff
@@ -152,14 +162,88 @@ public void testCombineUnderflow()
                 .negate()
                 .divide(BigInteger.valueOf(4));
 
-        assertEquals(average(state, TYPE), new BigDecimal(expectedAverage));
+        assertAverageEquals(expectedAverage);
+    }
+
+    @Test(dataProvider = "testNoOverflowIntegerDataProvider")
+    public void testNoOverflowInteger(List<BigInteger> numbers)
+    {
+        testNoOverflow(TYPE, numbers);
+    }
+
+    @Test(dataProvider = "testNoOverflowDecimalDataProvider")
+    public void testNoOverflowDecimal(List<BigInteger> numbers)
+    {
+        testNoOverflow(createDecimalType(38, 2), numbers);
+    }
+
+    private void testNoOverflow(DecimalType type, List<BigInteger> numbers)
+    {
+        for (BigInteger number : numbers) {
+            addToState(type, state, number);
+        }
+
+        assertEquals(state.getOverflow(), 0);
+        BigInteger sum = numbers.stream().reduce(BigInteger.ZERO, BigInteger::add);
+        assertEquals(getDecimalSlice(state), unscaledDecimal(sum));
+
+        BigDecimal expectedAverage = new BigDecimal(sum, type.getScale()).divide(BigDecimal.valueOf(numbers.size()), type.getScale(), HALF_UP);
+        assertEquals(decodeBigDecimal(type, average(state, type)), expectedAverage);
+    }
+
+    @DataProvider
+    public static Object[][] testNoOverflowIntegerDataProvider()
+    {
+        return new Object[][] {
+                {ImmutableList.of(TWO.pow(125), TWO.pow(124), ONE)},
+                {ImmutableList.of(TWO, ONE)},
+                {ImmutableList.of(ZERO, ONE)},
+                {ImmutableList.of(TWO.negate(), ONE.negate())},
+                {ImmutableList.of(ONE.negate(), ZERO)},
+                {ImmutableList.of(ONE.negate(), ZERO, ZERO)},
+                {ImmutableList.of(TWO.negate(), ZERO, ZERO)},
+                {ImmutableList.of(TWO.negate(), ZERO)}
+        };
+    }
+
+    @DataProvider
+    public static Object[][] testNoOverflowDecimalDataProvider()
```

Comment:
what's the difference between `testNoOverflowIntegerDataProvider` and `testNoOverflowDecimalDataProvider`?

`testNoOverflowDecimalDataProvider` seems to use small number too. Should it us numbers with actually need 38 precision?

Reply:
I merged the two providers into `testNoOverflowDataProvider` since test methods were merged into one.